sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/batch/BulkExecutor.java

```diff
@@ -0,0 +1,284 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.FluxProcessor;
+import reactor.core.publisher.FluxSink;
+import reactor.core.publisher.GroupedFlux;
+import reactor.core.publisher.Mono;
+import reactor.core.publisher.UnicastProcessor;
+import reactor.core.scheduler.Schedulers;
+
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static com.azure.cosmos.implementation.guava25.base.Preconditions.checkNotNull;
+
+public final class BulkExecutor<TContext> {
+
+    private final static Logger logger = LoggerFactory.getLogger(BulkExecutor.class);
+
+    private final CosmosAsyncContainer container;
+    private final AsyncDocumentClient docClientWrapper;
+    private final ThrottlingRetryOptions throttlingRetryOptions;
+    private final Flux<CosmosItemOperation> inputOperations;
+
+    // Options for bulk execution.
+    private final int maxMicroBatchSize;
+    private final int maxMicroBatchConcurrency;
+    private final Duration maxMicroBatchInterval;
+    private final TContext batchContext;
+
+    // Handle gone error:
+    private final AtomicBoolean mainSourceCompleted;
+    private final AtomicInteger totalCount;
+    private final FluxProcessor<CosmosItemOperation, CosmosItemOperation> mainFluxProcessor;
+    private final FluxSink<CosmosItemOperation> mainSink;
+    private final List<FluxSink<CosmosItemOperation>> groupSinks;
+
+    public BulkExecutor(CosmosAsyncContainer container,
+                        Flux<CosmosItemOperation> inputOperations,
+                        BulkProcessingOptions<TContext> bulkOptions) {
+
+        checkNotNull(container, "expected non-null container");
+        checkNotNull(inputOperations, "expected non-null inputOperations");
+        checkNotNull(bulkOptions, "expected non-null bulkOptions");
+
+        this.container = container;
+        this.inputOperations = inputOperations;
+        this.docClientWrapper = CosmosBridgeInternal.getAsyncDocumentClient(container.getDatabase());
+        this.throttlingRetryOptions = docClientWrapper.getConnectionPolicy().getThrottlingRetryOptions();
+
+        // Fill the option first, to make the BulkProcessingOptions immutable, as if accessed directly, we might get
+        // different values when a new group is created.
+        maxMicroBatchSize = bulkOptions.getMaxMicroBatchSize();
+        maxMicroBatchConcurrency = bulkOptions.getMaxMicroBatchConcurrency();
+        maxMicroBatchInterval = bulkOptions.getMaxMicroBatchInterval();
+        batchContext = bulkOptions.getBatchContext();
+
+        // Initialize sink for handling gone error.
+        mainSourceCompleted = new AtomicBoolean(false);
+        totalCount = new AtomicInteger(0);
+        mainFluxProcessor = UnicastProcessor.<CosmosItemOperation>create().serialize();
```

Comment:
This is a purely question for myself. 
When I check the UnicastProcessor documentation, it says it will be removed in 3.5. Should we change to something else just in case in the future we upgrade to > 3.5.0? Currently we are using 3.3.9
_"Deprecated. 
to be removed in 3.5, prefer clear cut usage of Sinks through variations under Sinks.many().unicast()."_

Reply:
As far as I know, they are just making it bit easy to use. SInk.Many I think will serve this purpose (i.e. multiple producer one consumer). New one also have complete and emit options, so should be few init changes I think when we bump up the version.

Reply:
https://github.com/reactor/reactor-core/blob/master/docs/asciidoc/processors.adoc

Reply:
I see, thanks for sharing the doc and sharing the details~